Utils/ClanUtils.cs

```diff
@@ -17,5 +50,212 @@ public static float RecalculateClanPP(this AppContext context, int clanId)
             }
             return resultPP;
         }
+
+        public static ICollection<ClanRanking> CalculateClanRanking(this AppContext context, Leaderboard leaderboard)
+        {
+            // CalculateClanRanking: Function that calculates the clanRanking given a leaderboard
+            // This function is called on relevant leaderboards whenever a user sets a new score, a clan is created, a user leaves a clan,
+            // a user joins a clan, a map is ranked, or a clan is deleted.
+            if (leaderboard == null) 
+            {
+                return null;
+            }
+
+            // Calculate owning clan on this leaderboard
+            var newClanRankingData = new Dictionary<Clan, ClanRankingData>();
+            var leaderboardClans =
+                context
+                    .Scores
+                    .Include(s => s.Player)
+                    .ThenInclude(p => p.Clans)
+                    .Where(s => s.LeaderboardId == leaderboard.Id && !s.Banned && s.Player.Clans.Count != 0)
+                    .OrderByDescending(el => el.Pp)
+                    .ToList();
+
+            // Build up a dictionary of the clans on this leaderboard based on scores
+            foreach (Score score in leaderboardClans)
+            {
+                if (score.Player.Clans == null)
+                {
+                    continue;
+                }
+                foreach (Clan clan in score.Player.Clans)
+                {
+                    if (newClanRankingData.ContainsKey(clan))
+                    {
+                        // Update the data already in the newClanRankingData dictionary
+                        newClanRankingData[clan].numberOfScores = newClanRankingData[clan].numberOfScores + 1;
+                        newClanRankingData[clan].weight = MathF.Pow(0.8f, newClanRankingData[clan].weight);
+                        newClanRankingData[clan].lastUpdateTime = Math.Max(Int32.Parse(score.Timeset), newClanRankingData[clan].lastUpdateTime);
+                        newClanRankingData[clan].clanPP = newClanRankingData[clan].clanPP + (score.Pp * newClanRankingData[clan].weight);
+                        newClanRankingData[clan].totalAcc = newClanRankingData[clan].totalAcc + score.Accuracy;
+                        newClanRankingData[clan].totalRank = newClanRankingData[clan].totalRank + score.Rank;
+                        newClanRankingData[clan].totalScore = newClanRankingData[clan].totalScore + score.ModifiedScore;
+                        newClanRankingData[clan].Scores.Add(score);
+                    }
+                    else
+                    {
+                        newClanRankingData.Add(clan, new ClanRankingData(
+                            1,
+                            1.0f,
+                            Int32.Parse(score.Timeset),
```

Comment:
I changed it to use timepost rather than timeset everywhere in this function. I hope that is appropriate, or did you mean "just this specific instance change to timepost".

Reply:
Yeap, changing everywhere is great. `timepost` is the timestamp of the server receiving score, `timeset` is the timestamp from replay for the recording started.